Add Timedones Plugin

mantikafasi opened the pull request:
- [x] Fix Timezones looking ugly in user profiles somehow (@Vendicated will figure out)
- [ ] Confirm This weird code is working
- [x] Bottom Text




- PS: Ignore commit history

mantikafasi committed:
make some changes in design and add badges to userreviews

mantikafasi committed:
delete bypassAutomod :trollface:

mantikafasi committed:
remove unneeded import trol

mantikafasi committed:
fix elint

mantikafasi committed:
MAKE USERNAMES CLICKABLE AGAIN

mantikafasi committed:
Merge branch 'main' of https://github.com/Vendicated/Vencord

mantikafasi committed:
Fix conflicts

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
remove "default" keyword from Badge because eslint explodes for some reaosn

mantikafasi committed:
make badges redirect using discords module

Vendicated committed:
Merge branch 'main' into main

mantikafasi committed:
eslint strikes again

mantikafasi committed:
fix stupit scrollbar showing at input if username includes emojsi

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'main' of https://github.com/Vendicated/Vencord

mantikafasi committed:
FİX

Vendicated committed:
Merge branch 'main' into main

mantikafasi committed:
Merge branch 'main' of https://github.com/Vendicated/Vencord

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'main' of https://github.com/mantikafasi/BetterVencord

mantikafasi committed:
Timezoes plugin

mantikafasi commented:
@AAGaming00 FİNİSH THİS

AAGaming00 commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,57 @@
+
+export interface Timezone {
+    userID: string,
+    timezoneId: String,
```
```suggestion
    timezoneId: string,
```
:trollface:

AAGaming00 reviewed (changes requested):
explode

mantikafasi committed:
Update Utils.ts

mantikafasi commented:
> explode

GUHHHHHHHHHHHHHH

AAGaming00 commented:
also you should probably type cache map

mantikafasi committed:
removed some questionable code

mantikafasi commented:
> also you should probably type cache map

I removed types entirely :trollface: 
I only store timezone now

AAGaming00 commented:
wdym what

AAGaming00 commented:
i mean ts types stupit

mantikafasi committed:
ruin code more

mantikafasi commented:
> i mean ts types stupit

fin

AAGaming00 commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+
+
+const API_URL = "https://timezonedb.catvibers.me/";
+const Cache = new Map<string, Number | null>();
```
```suggestion
const Cache = new Map<string, number | null>();
```
EXPLODE

mantikafasi replied:
I love "Number" more

AAGaming00 replied:
explode it's inconsistent

Vendicated replied:
Number is the class not the primitive, you should be using number

mantikafasi replied:
insane people everywhere

mantikafasi replied:
fine

ArjixWasTaken replied:
I wish js had int and float types

RGBCube replied:
> I wish js had int and float types

(he doesn't know) (BigInt) (Number)

ArjixWasTaken replied:
I do know

AAGaming00 reviewed (changes requested):
how do you do this twice

mantikafasi committed:
insn

C10udburst commented:
Timedones?

AAGaming00 commented:
> Timedones?

@mantikafasi guhhhh you are insane

AAGaming00 commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+        return Cache.get(discordID) as number | null;
+    }
+
+    const timezone = await fetch(API_URL + "api/user/" + discordID).then(
```
maybe type the fetch response with Timezone instead of any?

also you should .catch the fetch prob

abUwUser replied:
honestly, i don't like much of the use of TimezoneDB, mainly because of how that still relies on UTC offsets. If there is a different Timezone service out there, it could be a good idea.
(btw i am the og dev of timezonedb)

AAGaming00 replied:
> honestly, i don't like much of the use of TimezoneDB, mainly because of how that still relies on UTC offsets. If there is a different Timezone service out there, it could be a good idea.
> (btw i am the og dev of timezonedb)

this is new backend it doesn't work that way anymore stupit

it gives timezone now and client has to parse

AAGaming00 replied:
Actually this plugin uses the utc offset but I think it can do other way since backend provides I think

backend also generates correct offset anyway tho now I think

mantikafasi replied:
you will make api never go down so I wont have to

mantikafasi replied:
wait does backend provide timezone @AAGaming00

mantikafasi replied:
![image](https://user-images.githubusercontent.com/67705577/210106021-09b91b6f-845b-47c8-96ee-d3e5363422fb.png)
İT DOESNT

mantikafasi replied:
@rushiiMachine you will add

Vendicated replied:
it's right there wdym, America/New_York

mantikafasi replied:
ok I thought by timezone it would return something like "12:20" but that wouldnt be timezone whatever trol

AAGaming00 replied:
manti is 9000000 years old he has bad eyesight

mantikafasi replied:
yop

abUwUser replied:
> this is new backend it doesn't work that way anymore stupit
> 
> it gives timezone now and client has to parse

oh okay. i forgot about that change

AAGaming00 commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,186 @@
+    getProfileTimezonesComponent: (e: any) => {
+
+        if (!Vencord.Settings.plugins.Timezones.showTimezonesInProfile) {
+            return (<></>);
```
would null not make more sense here?

mantikafasi replied:
wouldnt returning null make react explod

AAGaming00 replied:
idk try it I'm pretty sure it doesn't in this scenario

AAGaming00 commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,186 @@
+                border: "4px solid white"
+            }}>
+                {timezone && getTimeString(timezone)}
+
```
dumb newline go explode immediately

mantikafasi replied:
OK GET CALM PLEASE

AAGaming00 commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,186 @@
+    }
+    ,
+    getTimezonesComponent: (e: any) => {
+        if (Vencord.Settings.plugins.showTimezonesInChat || e.user)
```
wtf is || e.user for shouldn't it be || !e.user

mantikafasi replied:
idk my past self wrote this, I have no idea why its like that

mantikafasi replied:
e.user doesnt even exist in that object normally :trollface: but i will leave it

AAGaming00 commented:
also run linter guhhhhh

mantikafasi commented:
> Timedones?

Timedones

abUwUser commented:
it could be a good idea to use city timezones (e.g: America/Sao_Paulo) mainly because of the city change. There is the [Temporal API proporsal](https://tc39.es/proposal-temporal/docs/), but it is a proposal, even tho it is a stage 3 one (likely to be adopted)

mantikafasi committed:
AAGAMING00 SO ANGRY

AAGaming00 commented:
> it could be a good idea to use city timezones (e.g: America/Sao_Paulo) mainly because of the city change. There is the [Temporal API proporsal](https://tc39.es/proposal-temporal/docs/), but it is a proposal, even tho it is a stage 3 one (likely to be adopted)

discord literally has moment just use it

mantikafasi commented:
> > it could be a good idea to use city timezones (e.g: America/Sao_Paulo) mainly because of the city change. There is the [Temporal API proporsal](https://tc39.es/proposal-temporal/docs/), but it is a proposal, even tho it is a stage 3 one (likely to be adopted)
> 
> discord literally has moment just use it

discord has moment but i think it doesnt have moment-timezone addon. and i think without it I cant parse timezone strings

ArjixWasTaken commented:
just add the addon

mantikafasi commented:
> just add the addon

ven will explode

ArjixWasTaken commented:
ven is dead, god is dead, manti is turkish

mantikafasi commented:
wrong (except the lat one )

somerandomcloud commented:
> manti is turkish

are you listing reasons why this pr should be rejected?

ArjixWasTaken commented:
> are you listing reasons why this pr should be rejected?

That was not my intention, but now that you say it, yes.

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+        Cache.set(discordID, null);
+        return null;
+    }
+    Cache.set(discordID, timezone.timezone);
```
You should use `timezone_id` which returns just an IANA timezone name, defaulting to the `timezone` offset when its not there.
Ref:
https://github.com/rushiiMachine/TimezoneDB/blob/9f6a4d0f1ea52c1a83d2cd18c5cd774727b452e1/src/routes/user.rs#L63-L67

C10udburst commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,182 @@
+
+    getProfileTimezonesComponent: (e: any) => {
+
+        if (!Vencord.Settings.plugins.Timezones.showTimezonesInProfile) {
+            return null;
+        }
```
why brackets??

C10udburst commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+    if (getSettings()[`timezones.${discordID}`])
+        return Number(getSettings()[`timezones.${discordID}`]);
+
+    if (Cache.has(discordID)) {
+        return Cache.get(discordID) as number | null;
+    }
```
why brackets????

Vendicated replied:
:trollface:

mantikafasi replied:
sometimes correct question is, why not

Vendicated commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+
+const API_URL = "https://timezonedb.catvibers.me/";
+const Cache = new Map<string, number | null>();
+export const moment: typeof import("moment") = findByPropsLazy("parseTwoDigitYear");
```
this is in webpack commons! (or should be at least)

Vendicated commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,54 @@
+        r => r.json()
+    );
+
+    if (timezone.error) {
```
why does the api return errors as okay responses?? or is this just the json body on 4xx/5xx? If it's the latter case, you should check the status code instead

rushiiMachine replied:
its the latter

Vendicated commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,182 @@
+    options: {
+        use24hr: {
+            type: OptionType.BOOLEAN,
+            description: "Use 24h format",
```
why is this a setting here? Please use Discord's locale settings along with Intl.DateTimeFormat
```js
const PreloadedUserSettings = findLazy(m => m.ProtoClass?.typeName === "discord_protos.discord_users.v1.PreloadedUserSettings");

const locale = PreloadedUserSettings .getCurrentValue().localization.locale.value;
new Intl.DateTimeFormat("en-US", { hour: "numeric", minute: "numeric" }).format(new Date)
```
![image](https://user-images.githubusercontent.com/45497981/210117961-fdd029c3-6030-44d5-99a1-8ecafd62acbc.png)

mantikafasi replied:
that would work but what if someone wanted to use with different format, like on here we use 24hr clock but I sometimes use 12hr on discord (even now)

mantikafasi replied:
buuut at the same time yeah probably no one would do this except few insane people

Vendicated replied:
Worsening the experience for the majority of people to cover a very minor use case is kinda dum

Vendicated replied:
Americans would cry for their mum if they see 24h time

mantikafasi replied:
true

mantikafasi committed:
less shitcode

ItzOnlyAnimal commented:
guhh explode

mantikafasi commented:
guhhhh very soon

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,177 @@
+                match: /(?<=return\s*\(0,\w{1,3}\.jsxs?\)\(.+!\w{1,3}&&)(\[{0,1}\(0,\w{1,3}.jsxs?\)\(.+?\{.+?\}\)*\]{0,1})/,
+                // DONT EVER ASK ME HOW THIS WORKS I DONT KNOW EITHER I STOLE IT FROM TYMEN
+                replace: "[$1, Vencord.Plugins.plugins.Timezones.getTimezonesComponent(e)]"
+            },
```
assuming `e` is from discord src why is it hardcoded

mantikafasi replied:
e is the parameter of discords function , which is always same

rushiiMachine replied:
fix
i dont know what that horrid regex is but fix

mantikafasi replied:
GUHHH ITS ALWAYS E THERE IS NO NEED TO FIX

rushiiMachine replied:
BAD WHAT IF IT CHANGES

mantikafasi replied:
IT WONT MINIFIER LITERALLY NAMES ALL FIRST PARAMETERS e
![image](https://user-images.githubusercontent.com/67705577/210159629-87dec5a6-de13-40b5-8674-ef0629e33fb2.png)

rushiiMachine replied:
you will explolde

ArjixWasTaken replied:
> IT WONT MINIFIER LITERALLY NAMES ALL FIRST PARAMETERS e

Minifier: *updates and decides to name all first parameters as f*
manti: guh

ArjixWasTaken replied:
if it is e now, it does not mean that it will be e in the future

mantikafasi replied:
guhh its been e for a long time, I dont think it will ever change

AAGaming00 replied:
you are e

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,177 @@
+            replacement: {
+                match: /return(\(0,.\.jsx\)\(.\..,\{.*\}\))/,
+                replace: "return [$1, Vencord.Plugins.plugins.Timezones.getProfileTimezonesComponent(e)]"
+            }
```
here too

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,52 @@
+        return Cache.get(discordID) as string | null;
+    }
+
+    const response = await fetch(API_URL + "api/user/" + discordID);
```
valid user agent header here
`Vencord/1.0.0` or something idk

ArjixWasTaken replied:
Imo user agent doesn't mean much, cause that can be faked as easily as breathing.

rushiiMachine replied:
not the point its analytics for aa

Vendicated replied:
it's not possible to send a custom user agent

Vendicated replied:
https://bugs.chromium.org/p/chromium/issues/detail?id=571722

AAGaming00 replied:
X-User-Agent :trolley:

rushiiMachine replied:
yeah or that works too

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,177 @@
+    description: "Shows the timezones of users",
+    authors: [Devs.mantikafasi],
+    options: {
+
```
can you add a button link to `https://timezonedb.catvibers.me?client_mod=Vencord` on the settings page
"Set own public Timezone" or whatever

Vendicated replied:
why, just do it in client

rushiiMachine replied:
manti will mess the login up and it will break
just link to external it works perfectly fine

Vendicated replied:
poor ux, just do it the same way as the reviewdb plugin

mantikafasi replied:
we will use webview

AAGaming00 replied:
DO NOT USE WEBVIEW GUHHHHH

mantikafasi replied:
you love

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,52 @@
+
+const getSettings = () => Vencord.Settings.plugins.Timezones;
+
+export async function getUserTimezone(discordID: string): Promise<string | null> {
```
also batch fetching is really needed, say every 5 seconds or whatever
currently it ddos's AA instead everytime you open a chat
with like 500 duplicate requests

use the new [bulk route](https://github.com/rushiiMachine/TimezoneDB#post-apiuserbulk) i added

Vendicated replied:
you should just take the fetching logic from pronoundb

mantikafasi replied:
@Vendicated you will do this, I hate stupit promises

AAGaming00 replied:
> @Vendicated you will do this, I hate stupit promises

explode

ItzOnlyAnimal replied:
you will do this for aliucord plugin too @mantikafasi :trollface:

mantikafasi replied:
soon

mantikafasi replied:
I will refuse to do this

AAGaming00 replied:
> I will refuse to do this

NO

mantikafasi committed:
make it ignore current user

mantikafasi committed:
better

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -130,12 +130,12 @@ export default definePlugin({
-        if (!Vencord.Settings.plugins.Timezones.showTimezonesInProfile) {
+        const user = e.user as User;
+
+        if (!Vencord.Settings.plugins.Timezones.showTimezonesInProfile || user.id === UserStore.getCurrentUser().id) {
```
make it show for youself too so u can tell if it's off or not

mantikafasi replied:
okey

mantikafasi committed:
fulfill what rusher wants

KraXen72 commented:
average mantika pr: 103 comments and still not merged

mantikafasi commented:
Because ven still not finished
@Vendicated FINISH

AAGaming00 commented:
@mantikafasi YOU WILL FINISH NOW

rushiiMachine commented:
@mantikafasi YOU WILL FINISH NOW

dzshn commented:
@mantikafasi YOU WILL FINISH NOW

ActuallyTheSun commented:
we will get to 4 digit comments before its finished

mantikafasi commented:
ven will finish

RGBCube commented:
finish yourself lazy, you have half a year of no school

mantikafasi commented:
@RGBCube I see this as a great opportunity to learn typescript for you

rushiiMachine commented:
add bulk fetching 👎

mantikafasi commented:
@rushiiMachine

ArjixWasTaken commented:
I'll share the code from my, now abandoned, timezones plugin

```typescript
import { Devs } from "@utils/constants";
import { useForceUpdater } from "@utils/misc";
import definePlugin from "@utils/types";
import { findByCodeLazy, findByPropsLazy } from "@webpack";
import { Text, useEffect, useState } from "@webpack/common";

import { timezones } from "./all_timezones";

const EditIcon = findByCodeLazy("M19.2929 9.8299L19.9409 9.18278C21.353 7.77064 21.353 5.47197 19.9409");
const classNames = findByPropsLazy("customStatusSection");

const RenderUserTimezone = ({ user: { id: userId } }) => {
    const [timeZone, setTimeZone] = useState<string | undefined>(undefined);
    const forceUpdate = useForceUpdater();

    useEffect(() => {
        setTimeout(forceUpdate, 1000);
        timezones; // do smth with this
    });

    return (
        <Text variant="text-sm/normal" className={classNames.customStatusSection}>
            {/* TODO: Make the icon and text to be vertically centered. */}
            {(new Date).toLocaleString("en-uk", { dateStyle: "long", timeStyle: "short", timeZone }).replace(" at", ",")}
            <span
                style={{ cursor: "pointer" }}
                onClick={() => {
                    console.log("ayy!");
                }}
            >
                <EditIcon color="var(--primary-light-530)" height="18" width="18" />
            </span>
        </Text>
    );
};


export default definePlugin({
    name: "UserTimezones",
    description: "Allows you to pick a timezone for each user to show up in their user modal.",
    authors: [Devs.Arjix],
    patches: [{
        find: "().customStatusSection",
        replacement: {
            // Inserts the timezone component right below the custom status.
            match: /user:(\w),nickname:\w,.*?children.*?\(\)\.customStatusSection.*?\}\),/,
            replace: "$&$self.renderTimezone({user:$1}),"
        }
    }],
    renderTimezone: ({ user }) => <RenderUserTimezone user={user} />,
});
```

I am sharing this bc of
![image](https://user-images.githubusercontent.com/53124886/219854772-166159d6-1739-443e-ad1e-117fca5b1e9f.png)

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
update profile component

mantikafasi commented:
![image](https://user-images.githubusercontent.com/67705577/219857266-e7bbb502-d1d4-4cb3-91d4-5d8165147d1f.png)
tbh I was thinking banner approach is better but this looks pretty good and clean too, ty

AAGaming00 commented:
YOU WILL ADD BATCH FETCH

ArjixWasTaken committed:
Migrated to DataStore

ArjixWasTaken committed:
Bulk fetch timezones to not spam the API

Although this doesn't currently work as CORS is fucking us over.

ArjixWasTaken commented:
Waiting for the timezones API to resolve the CORS issue, or if I am doing smth wrong to be pointed out.

ItzOnlyAnimal commented:
guhzones also add bulk fetch NOW

ArjixWasTaken commented:
I added bulk fetch...

ArjixWasTaken committed:
Update index.tsx

AAGaming00 commented:
> I added bulk fetch...

cors is fixed i think idk you will test NOW

AAGaming00 commented:
> guhzones also add bulk fetch NOW

yop manti so lazy

ArjixWasTaken committed:
feat: edit the timezone of a user

for now this is just temporarily changing the timezone, nothing gets changed in the DB
just for testing purposes

ArjixWasTaken committed:
Save the custom timezone and handle edge cases.

ArjixWasTaken committed:
Add a custom about section that also mentions TimezoneDB.

ArjixWasTaken commented:
I can confirm that this weird code works and that it is no longer a draft, but it is ready for reviews.

Currently I already know the following:
- I should refactor the inline styles to be in classes instead
- The in-message "local" timestamp doesn't update properly

ArjixWasTaken commented:
## Screenshots:

### User Modal
![image](https://user-images.githubusercontent.com/53124886/219954154-77c6bf9c-a653-4e21-9ac2-80dfaa8abbf1.png)
![image](https://user-images.githubusercontent.com/53124886/219954175-111adb30-9677-448c-bd2b-2b0eef8f4e0d.png)
![image](https://user-images.githubusercontent.com/53124886/219954187-40fcbcb4-ac34-45b4-a221-7f3a8ff66f81.png)

### Message
![image](https://user-images.githubusercontent.com/53124886/219954263-8eeb8c78-d244-4fab-9fe9-cd84f4d7bf47.png)

ArjixWasTaken committed:
Update list of all timezones to include all the timezones from TimezoneDB

ArjixWasTaken committed:
Fixed the user agent header

ArjixWasTaken committed:
a commit name

mantikafasi committed:
Update index.tsx

rushiiMachine commented on src/plugins/Timezones/all_timezones.ts:
```diff
@@ -0,0 +1,399 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+export const timezones = [
```
why
just include the package you copied this from
@ultirequiem/timezones

Vendicated replied:
or just don't add 10kb of timezones wtf

ArjixWasTaken replied:
> just include the package you copied this from

uhhh......I did not use a package for this shit...
Initially I downloaded the data from [iana](https://www.iana.org/time-zones) and wrote a python script to extract all the timezones from all the files

but the TimezoneDB does not use timezone codes, rather it uses city names, so I scraped the timezones from TimezoneDB, and I included the classic Etc/GMT timezones from iana

ArjixWasTaken replied:
so in fact, this list is semi-manually made

ArjixWasTaken replied:
[here](https://github.com/Vendicated/Vencord/pull/376/commits/0d656f5e00ef6e20d83b3dfd7d5c4503aab29925#diff-ac6b600d7a21ce01d06c26ec87b50c60851b19120aa31c27d971f892b325e9d3) is the original list, with the timezones exclusively from iana

ItzOnlyAnimal replied:
:trollface:

mantikafasi commented:
guys please its just timezone plugin dont make this most commented issue on github

Vresod commented:
Can't I'm busy making this the most commented pull request on github

Captain8771 commented:
> Can't I'm busy making this the most commented pull request on github

I hope you explode

mantikafasi commented:
> > > Can't I'm busy making this the most commented pull request on github
> > 
> > 
> > I hope you explode
> 
> I hope you explode

I hope you explode

ArjixWasTaken commented:
> > > > Can't I'm busy making this the most commented pull request on github
> > > 
> > > 
> > > I hope you explode
> > 
> > I hope you explode
> 
> I hope you explode

I hope you defenestrate yourself

Vresod commented:
> > > > > Can't I'm busy making this the most commented pull request on github
> > > > 
> > > > 
> > > > I hope you explode
> > > 
> > > 
> > > I hope you explode
> > 
> > 
> > I hope you explode
> 
> I hope you defenestrate yourself

I hope you explode

Vendicated commented:
stop spamming

mantikafasi commented:
merge and we will stop

mantikafasi committed:
Merge branch 'Vendicated:main' into main

ItzOnlyAnimal commented:
**>** yop


https://user-images.githubusercontent.com/24845294/221240941-779e7028-699e-4396-84cb-97cfa4b3ed88.mp4

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'Vendicated:main' into main

abUwUser commented:
is this ready for use as of now?

mantikafasi commented:
I think its ready to use but need to make some small changes. we need to change select component into the one that supports typing with autocomplete. We also need to make it fetch timezone strings from somewhere instead of hardcoding json

lewisakura commented:
> we need to change select component into the one that supports typing with autocomplete.

now possible due to #518, shouldn't be too difficult to convert.

> We also need to make it fetch timezone strings from somewhere instead of hardcoding json

`Intl.supportedValuesOf('timeZone');` for electron 17, but until stable reaches that version I have a copy of what it returns [hosted on a gist](https://gist.githubusercontent.com/lewisakura/529bb0236f375ab0b5890de02a8c007a/raw/ff5930de4d8b6a77e32bba3d1f8b822a80286493/timezones.json). you can use this or rehost it somewhere, or go for a completely different source.

personally i'd implement it like so:
```typescript
async getTimezones() {
    return Intl.supportedValuesOf?.("timeZone") ?? await fetch("...").then(r => r.json());
}
```
but i'll leave implementation detail up to you

mantikafasi committed:
switch to SearchableSelect

4dvyte commented:
So...is this going to be a thing?

ArjixWasTaken commented:
working on it rn

mantikafasi commented:
it will be within 6 months

AAGaming00 commented:
10 year old pr soon

ArjixWasTaken committed:
The end

ArjixWasTaken committed:
bye bye

ArjixWasTaken committed:
Merge branch 'main' into main

ArjixWasTaken commented:
Done

ArjixWasTaken committed:
fix: Linter issues

mantikafasi commented:
check fail

AAGaming00 commented:
you are a fail

mantikafasi commented:
also why did u yeet slash commands

ArjixWasTaken commented:
> also why did u yeet slash commands

you yourself said you didn't really want them anymore, I was the one that said that it's ok to keep them

ArjixWasTaken committed:
fix: I hate this linter

mantikafasi commented:
yes but later I think we disscussed about it and thought it would be good if they stayed. Since people might wanna disable tz component on user profiles.

But whatever people will just not disable tz component on profiles :trollface: 

Also I think we discussed to add some slash command that sends "To get your timezone please visit blabla website" to get other persons timezone string

ArjixWasTaken commented:
I hate my life
![image](https://user-images.githubusercontent.com/53124886/230797061-4839648d-13c3-4c18-b308-4b14e5389b11.png)

mantikafasi commented:
:salute:

ArjixWasTaken committed:
minor tweak

ArjixWasTaken commented:
This should be done, @Vendicated review when?

ArjixWasTaken commented:
> Also I think we discussed to add some slash command that sends "To get your timezone please visit blabla website" to get other persons timezone string

Personally I am not really a fan of that idea, you can add it if you want.

mantikafasi commented:
I also think its not very useful but lewi already made a website so we can just add it

mantikafasi commented:
also @Vendicated review

ArjixWasTaken commented:
> I also think its not very useful but lewi already made a website so we can just add it

ohhhh, you are talking about *that*, I thought you were talking about a list of all the timezones

mantikafasi commented:
yop add and ven will merge

ArjixWasTaken commented:
You add it, I'm off to sleep.

mantikafasi commented:
guhhh

mantikafasi commented:
I dont even have this prs repo on my pc anymore trol you do this tomorrow

ArjixWasTaken commented:
> I dont even have this prs repo on my pc anymore trol you do this tomorrow

why is that an excuse?

ArjixWasTaken commented:
Also I don't have the link, and I am not in the discord server.

mantikafasi commented:
> > I dont even have this prs repo on my pc anymore trol you do this tomorrow
> 
> why is that an excuse?

I am just lazy to git pull and change vencords location for one simple thing, its like 3 lines of code otherwise h. I am not in server either but heres url https://gh.lewisakura.moe/timezone/

mantikafasi committed:
add slash command and small error handling

mantikafasi commented:
after 4 and half months, I think this is ready to merge

mantikafasi committed:
I dont like this name

AAGaming00 commented:
> after 4 and half months, I think this is ready to merge

no it isnt

mantikafasi commented:
> > after 4 and half months, I think this is ready to merge
> 
> no it isnt

It is

mantikafasi committed:
Merge branch 'Vendicated:main' into main

TheColorman commented:
What are the timezones based on? I noticed a lot of example cities missing (e.g. Montgomery) and am wondering whether there is a reason for that.

ArjixWasTaken commented:
The timezones are taken from Intl, a JavaScript built-in library.

One possible reason for a city missing is that multiple cities can have the same timezone and there is no point having many cities.

For example Greece has only one timezone, there is no reason to have more cities than Athens.

But yeah, Intl is where we are getting the timezones from.

(If you are talking about the website TimezoneDB then I don't know)

4dvyte commented:
The plugin is functional but the time does not show up on the profile everywhere.

shows here
![image](https://user-images.githubusercontent.com/104263903/235625641-376028dc-21c4-4fcb-9846-5c81b0bd63fd.png)

but does not show up here.
![image](https://user-images.githubusercontent.com/104263903/235625744-b0ba355e-a1c2-40cc-936f-a64124626ad0.png)
![image](https://user-images.githubusercontent.com/104263903/235625760-77fc98b3-2368-4528-a32d-0e2e05ab923f.png)

mantikafasi commented:
1: it only works in profile popout modal
2: utc offsets are kinda horrible because of summer time etc, they will be out of sync at some times

4dvyte commented:
Also what about the option to add 24 hour time?

ArjixWasTaken committed:
Merge branch 'main' into main

ArjixWasTaken committed:
ven dum dum

mantikafasi commented:
merge when

TheKodeToad commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,256 @@
+    description: "Allows you to see and set the timezones of other users.",
+    authors: [Devs.mantikafasi, Devs.Arjix],
+
+    commands: [
+        {
+            name: "timezone",
+            description: "Sends link to a website that shows timezone string, useful if you want to know your friends timezone",
+            execute: () => {
+                return { content: "https://gh.lewisakura.moe/timezone/" };
+            }
+        }
+    ],
```
This is kind of weird. First of all, the url should *obviously* be `https://gh.lewisakura.moe/timedone/` :trollface: Secondly, is this needed..?

Sqaaakoi replied:
what is the point of this???????? it has been marked as "resolved" 9 months ago and almost nothing has changed and it still seems useless

chocolatederp commented:
could you add something that allows you to manually set the timezone? like for example you put in +6 then its UTC+6, kinda like what the betterdiscord timezone plugin does, this is mainly because of the fact that some countries just arent available in the dropdown menu thats used currently

Vendicated commented:
go here to get ur timezone https://gh.lewisakura.moe/timezone/

TheColorman commented:
This has been a while in the making, any reason it can't be merged now and updated later? Most of the discussions and feature requests have been nice-to-haves, not really essentials for the function of the plugin. I see no reason not to merge.

mantikafasi commented:
I agree

Advyt3 commented:
So is this going to be a thing? or....

resucutie commented:
Feel free to remove me as reviewer. I do no longer work with timezone related stuff

Vendicated commented:
@CodiumAI-Agent /review

CodiumAI-Agent commented:
## PR Review

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

4, due to the complexity and size of the PR, which includes multiple new features and modifications across different files. The PR introduces a new plugin with various functionalities, including timezone fetching, UI components for displaying and editing timezones, and integration with external APIs. Reviewing this PR requires a thorough understanding of the existing codebase, the new functionalities being added, and potential impacts on performance and user experience.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td rowspan=2> 🔍&nbsp;<strong>Possible issues</strong></td>
<td>

<strong>Possible Bug:</strong> The `bulkFetchTimezones` function does not handle the case where the `fetch` request fails due to network issues or the server being down. This could lead to unhandled promise rejections.</td></tr>
<tr>
<td>

<strong>Performance Concern:</strong> The `getUserTimezone` function makes asynchronous calls within a `Promise` constructor but does not handle errors that may occur during these asynchronous operations. This could lead to unhandled promise rejections and potentially impact the user experience.</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>src/plugins/Timezones/Utils.ts
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider adding error handling for the `fetch` request within the `bulkFetchTimezones` function to gracefully handle cases where the request fails. This could include retrying the request a certain number of times before giving up or providing a fallback mechanism. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Vendicated/Vencord/pull/376/files#diff-596ef0efb45abace3a2f08bc0210d647160624197f7f633fd5e0fa71a1ce001aR51'>const req = await fetch(`${API_URL}/api/user/bulk`, {</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/plugins/Timezones/Utils.ts
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Implement error handling for the asynchronous operations within the `getUserTimezone` function. This could involve adding `.catch` blocks to the promises or using try/catch blocks within an async function to ensure that errors are caught and handled appropriately. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Vendicated/Vencord/pull/376/files#diff-596ef0efb45abace3a2f08bc0210d647160624197f7f633fd5e0fa71a1ce001aR96'>const timezone = (DataStore.get(DATASTORE_KEY) as Promise<TimezoneDB | undefined>).then(tzs => tzs?.[discordID]);</a></td></tr></table><hr>

</details><hr>

<details> <summary><strong>✨ Review tool usage guide:</strong></summary><hr> 

**Overview:**
The `review` tool scans the PR code changes, and generates a PR review which includes several types of feedbacks, such as possible PR issues, security threats and relevant test in the PR. More feedbacks can be [added](https://pr-agent-docs.codium.ai/tools/review/#general-configurations) by configuring the tool.

The tool can be triggered [automatically](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) every time a new PR is opened, or can be invoked manually by commenting on any PR.
- When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L23) related to the review tool (`pr_reviewer` section), use the following template:
```
/review --pr_reviewer.some_config1=... --pr_reviewer.some_config2=...
```
- With a [configuration file](https://pr-agent-docs.codium.ai/usage-guide/configuration_options/), use the following template:
```
[pr_reviewer]
some_config1=...
some_config2=...
```
    

See the review [usage page](https://pr-agent-docs.codium.ai/tools/review/) for a comprehensive guide on using this tool.


</details>

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
fix timezonedb

mantikafasi committed:
fix linter

mantikafasi committed:
die tsc

Variable1485 commented:
Is this still going?

mantikafasi commented:
yes

nin0-dev commented:
One day, this will get merged. Vencord will already have been discontinued in favor of Cumcord v2, Manti kafasi will be a grown man with 4 wives and 10 children, and will no longer have time to code in Java. Windows 10 will have been discontinued, the inflation rate will have gone up by 300%.

Variable1485 commented:
> One day, when this will get merged, Vencord will already have been discontinued in favor of Cumcord v2, Manti kafasi will be a grown man with 4 wives and 10 children, and will no longer have time to code in Java. Windows 10 will have been discontinued, the inflation rate will have gone up by 300%, and DevilBro will have taken over the Electron development space.

so true

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'main' into main

mantikafasi commented:
appearently this is a bit broken rn

AAGaming00 reviewed (changes requested):
i suggest you delete this horrible code and then delete your pc to prevent this from ever happening again @mantikafasi

mantikafasi commented:
> i suggest you delete this horrible code and then delete your pc to prevent this from ever happening again @mantikafasi

YOU WANTED ME TO MAKE THIS PLUGIN WHY DO YOU TELL ME TO DELETE IT

Sqaaakoi commented:
> i suggest you delete this horrible code and then delete your pc to prevent this from ever happening again @mantikafasi

Have you tried bombing FakeNitro instead :trollface:

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,121 @@
+    }
+});
+
+export function getUserTimezone(discordID: string, strategy: CustomTimezonePreference):
+    Promise<string | undefined> {
+
+    return new Promise(res => {
+        const timezone = (DataStore.get(DATASTORE_KEY) as Promise<TimezoneDB | undefined>).then(tzs => tzs?.[discordID]);
+        timezone.then(tz => {
```
why not async?

mantikafasi replied:
IDK

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,121 @@
+    });
+}
+
+const gist = "e321f856f98676505efb90aad82feff1";
+const revision = "91034ee32eff93a7cb62d10702f6b1d01e0309e6";
+const timezonesLink = `https://gist.githubusercontent.com/ArjixWasTaken/${gist}/raw/${revision}/timezones.json`;
```
just make this into one constant lol

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,247 @@
+    commands: [
+        {
+            name: "timezone",
+            description: "Sends link to a website that shows timezone string, useful if you want to know your friends timezone",
```
```suggestion
            description: "Sends a link to a utility website that shows your current timezone identifier",
```

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,247 @@
+        return (
+            <Text variant="text-md/normal">
+                A plugin that displays the local time for specific users using their timezone. <br />
+                Timezones can either be set manually or fetched automatically from the <a href={href}>TimezoneDB</a>
```
```suggestion
                Timezones can either be set manually or fetched automatically from <a href={href}>TimezoneDB</a>.
```

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,247 @@
+            <span className={classes(styles.timestampInline, styles.timestamp)}>
+                {
+                    timezone && "• " + getTimeString(timezone,
+                        /* message.timestamp is actually Date but as discord-types is outdated I had to do this */
+                        ((message.timestamp as unknown) as Date))
```
is this still the case

nyakowint replied:
probably lol discord-types so horror

Sqaaakoi replied:
i've ran into a comment in discord-types that reads `@returns This return object is fucking huge; I'll type it later.` so I wouldn't be surprised if that isn't the case lmao

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,121 @@
+        });
+
+        return await req.json()
+            .then((res: { [userId: string]: { timezoneId: string; } | null; }) => {
+                const tzs = (Object.keys(res).map(userId => {
+                    return res[userId] && { [userId]: res[userId]!.timezoneId };
+                }).filter(Boolean) as TimezoneDB[]).reduce((acc, cur) => ({ ...acc, ...cur }), {});
```
PLEASE CLEAN THIS UP

nyakowint replied:
mantika you will be mapped filtered and as null and reduced to atoms if you dont explode this

mantikafasi replied:
I cant read this anymore tbh

Sqaaakoi replied:
> I cant read this anymore tbh

I don't want to even try to make sense of this

AAGaming00 replied:
manti summoning all of hell itself to write his code

rushiiMachine commented on src/plugins/Timezones/Utils.ts:
```diff
@@ -0,0 +1,121 @@
+    const ids = Object.keys(requestQueue);
+    const timezones = await bulkFetchTimezones(ids);
+    if (!timezones) {
+        // retry after 15 seconds
+        setTimeout(bulkFetch, 15000);
```
There should be a max retry count of like 3-5 to prevent ddos'ing aa's network if it goes down.

AAGaming00 replied:
timezonedb ddos incident #15

rushiiMachine commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,247 @@
+        }
+    ],
+
+    settingsAboutComponent: () => {
```
I would love if this has integration with the tzdb api to set your own timezone publicly from within the settings menu. I don't think oauth login would be too hard to implement, you've done it before in rdb

mantikafasi replied:
too hard

nyakowint replied:
nuh uh

rushiiMachine replied:
you are so lazy

Sqaaakoi replied:
+1 manti lazy

Sqaaakoi commented on src/plugins/Timezones/index.tsx:
```diff
@@ -0,0 +1,247 @@
+            getUserTimezone(user.id, preference).then(tz => setTimezone(tz));
+
+            // Rerender every 10 seconds to stay in sync.
+            const interval = setInterval(forceUpdate, 10 * 1000);
```
horror

mantikafasi replied:
if you are a person like ven who opens someones profile and stares at banner for 15 minutes it will display wrong content

AAGaming00 replied:
as long as it doesn't spam api ig it's fine but still doesn't discord have a proper component for live updating timestamp

Vendicated replied:
yes, it's used by ShowTimeoutDuration plugin

Vendicated replied:
actually im stupid that's for countdown

but yeah timestamp too, you can just `Parser.parse("<:t:12345678>")`

Sqaaakoi reviewed (commented):
:alien:

mantikafasi commented:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

















```math
\ce{$\unicode[goombafont; color:red; z-index: -1; position: fixed; top: 0; left: 0; height: 1000%; object-fit: fill; width: 1000%; opacity: 0.7; background-size:cover; background: url('https://github.com/Vendicated/Vencord/assets/67705577/ca775680-f208-4c1f-90a5-679ebf764f1b');]{x0000}$}
```

byeoon commented:
hmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm



$$\ce{$unicode[goombafont; color:blue; z-index: -1; position: fixed; top: 0; left: 0; height: 1000%; object-fit: fill; width: 1000%; opacity: 0.9; background-size: cover; background: url('https://cdn.discordapp.com/emojis/1236484138304012418.png?size=256&quality=lossless');]
{x0000}$}$$

byeoon commented:
testaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa





$$\ce{$\unicode[goombafont; color:red; z-index: -1; position: fixed; top: 0; left: 0; height: 1000%; object-fit: fill; width: 1000%; opacity: 0.9; background-size: cover; background: url('https://cdn.discordapp.com/emojis/1236484138304012418.png?size=256&quality=lossless');]
{x0000}$}$$

KrstlSkll69 commented:
Erm what the sigma @byeoon

byeoon commented:
> Erm what the sigma @byeoon

eheheheheheheheheheeehe

roxwize commented:
what in the world

nin0-dev commented:
> 
> https://github.com/Vendicated/Vencord/assets/91651232/377ad42d-2efc-4e42-a828-aafa78dd4644
> 
> 

what?

roxwize commented:
is it even worth trying to make a local time plugin with this existing

mantikafasi committed:
add special thakns

ArjixWasTaken committed:
Merge branch 'main' into main

ArjixWasTaken committed:
Revert "add special thakns"

This reverts commit a0b1e2162c1f0cb9b54a28763234cc28f1d779d8.

ArjixWasTaken committed:
add special thanks

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'Vendicated:main' into main

mantikafasi committed:
Merge branch 'Vendicated:main' into main

Vendicated commented:
This PR will be merged once it turns old enough to use Discord

mantikafasi commented:
which means it will be merged eventually

Hazrtine commented:
3942 laying timedones to rest vro

mantikafasi committed:
Merge branch 'Vendicated:main' into main

gaxolotl commented:
why does this still exist

ArjixWasTaken committed:
Merge branch 'main' into main

OllyNicholass commented:
Anyone picking this up? Would be a nice addition
